tests/test_bosonic_solver.py

```diff
@@ -3,23 +3,157 @@
+            [N_cut + 1] * kcut, N_cut
+        )
+        
+    total_nhe = int(
+            factorial(N_cut + kcut)
+            / (factorial(N_cut) * factorial(kcut))
+        )
+    assert nhe, total_nhe
+        
+def test_check_H():
+    """Tests the function for checking system Hamiltonian"""
+    assert(_check_Hsys(sigmax()))
+    assert(_check_Hsys([sigmax(), sigmaz()]))
+    assert(_check_Hsys([[sigmax(), np.sin], [sigmaz(), np.cos]]))
+    assert(_check_Hsys([[sigmax(), np.sin], [sigmaz(), np.cos]]))
+    assert(_check_Hsys(QobjEvo([sigmaz(), sigmax(), sigmaz()])))
 
-from bofin.heom import add_at_idx
 
+    err_msg = r"Hamiltonian format is incorrect."
 
-def test_add_at_idx():
+    with pytest.raises(RuntimeError, match=err_msg):
+       _check_Hsys(sigmax().full())
+
+    with pytest.raises(RuntimeError, match=err_msg):
+       _check_Hsys([[1 , 0], [0, 1]])
+
+    with pytest.raises(RuntimeError, match=err_msg):
+       _check_Hsys([sigmax(), [[1 , 0], [0, 1]]])
+
+    err_msg = r"Incorrect time dependent function for Hamiltonian."
+
+    with pytest.raises(RuntimeError, match=err_msg):
+       _check_Hsys([[sigmax(), [0, np.pi]]])
+
+    with pytest.raises(RuntimeError, match=err_msg):
+       _check_Hsys([[sigmax(), np.sin(0.5)]])
+
+
+@pytest.mark.filterwarnings("ignore::scipy.integrate.IntegrationWarning")
+@pytest.mark.parametrize(['bnd_cut_approx', 'tol'], [
+    pytest.param(True, 1e-4, id="bnd_cut_approx"),
+    pytest.param(False,  1e-3, id="no_bnd_cut_approx"),  
+])
+def test_pure_dephasing_model_HSolverDL( bnd_cut_approx,  tol):
     """
-    Tests the function to add at hierarchy index.
+    HSolverDL: Compare with pure-dephasing analytical assert that the
+    analytical result and HEOM produce the same time dephasing evoltion.
     """
-    seq = (2, 3, 4)
-    assert (add_at_idx(seq, 2, 1), (2, 3, 5))
+    cut_frequency = 0.05
+    coupling_strength = 0.025
+    lam_c = coupling_strength / np.pi
+    temperature = 1 / 0.95
+    times = np.linspace(0, 10, 21)
 
-    seq = (2, 3, 4)
-    assert (add_at_idx(seq, 0, -1), (1, 3, 4))
+    def _integrand(omega, t):
+        J = 2*lam_c * omega * cut_frequency / (omega**2 + cut_frequency**2)
+        return (-4 * J * (1 - np.cos(omega*t))
+                / (np.tanh(0.5*omega / temperature) * omega**2))
+
+    # Calculate the analytical results by numerical integration
+    expected = [0.5*np.exp(quad(_integrand, 0, np.inf, args=(t,))[0])
+                for t in times]
+
+    H_sys = Qobj(np.zeros((2, 2)))
+    Q = sigmaz()
+    initial_state = 0.5*Qobj(np.ones((2, 2)))
+    projector = basis(2, 0) * basis(2, 1).dag()
+    options = Options(nsteps=15_000, store_states=True)
+    hsolver = HSolverDL(H_sys, Q, coupling_strength, temperature,
+                        14, 2, cut_frequency,
+                        bnd_cut_approx=bnd_cut_approx,
+                        options=options)
+    test = expect(hsolver.run(initial_state, times).states, projector)
```

Comment:
Since we're testing the solver results we should probably assert things about the rest of the result too:
```
results = hsolver.run(...)
expected = expect(result.states, projector)
np.testing.assert_allclose(test, expected, atol=tol)
... assert things about other parts of the result ...
```

Reply:
I guess we can just check the output is normalized at least, there's not much else we can do with this particular example I think.